### PR TITLE
Attempt to fix attributes dropdown

### PR DESCRIPTION
**Pull request by petrican**

### CODE REVIEW

**Comment by petrican** [comment]
https://www.loom.com/share/73078f95f0c04780a228948adb5f595a

This is an attempt to fix the dropdown and in the same time preserve the functionality with the dropup in modal.

File: frontend/src/components/widget/Attributes/AttributesDropdown.js [hunk 1]
```diff
@@ -161,6 +177,7 @@ AttributesDropdown.propTypes = {
   handlePatch: PropTypes.func.isRequired,
   disableOnClickOutside: PropTypes.func.isRequired,
   enableOnClickOutside: PropTypes.func.isRequired,
+  arrIndex: PropTypes.number,
```
Comment (teosarca) [inline frontend/src/components/widget/Attributes/AttributesDropdown.js:180]: pls document what's `arrIndex`?

File: frontend/src/components/widget/Attributes/AttributesDropdown.js [hunk 2]
```diff
@@ -123,10 +130,19 @@ class AttributesDropdown extends PureComponent {
...
+
     return (
       <div
         ref={this.selector}
-        className="attributes-dropdown panel-shadowed panel-primary panel-bordered panel-spaced"
+        className={classnames(
+          'attributes-dropdown panel-shadowed panel-primary panel-bordered panel-spaced',
+          { 'hidden ': !isVisible },
+          {
+            'attributes-dropup': arrIndex > 13,
```
Comment (teosarca) [inline frontend/src/components/widget/Attributes/AttributesDropdown.js:143]: extract 13 to a constant. pls document what's that.

File: frontend/src/components/widget/Attributes/Attributes.js [hunk 3]
```diff
@@ -118,8 +118,11 @@ export default class Attributes extends Component {
    * @todo Write the documentation
    */
   handleToggle = (option) => {
-    const { handleBackdropLock } = this.props;
-    const { loading } = this.state;
+    const { handleBackdropLock, updateHeight, arrIndex, isModal } = this.props;
+    const { loading, dropdown } = this.state;
+
+    !dropdown && !isModal && arrIndex < 13 && updateHeight(400);
+    dropdown && !isModal && arrIndex < 13 && updateHeight(150);
```
Comment (teosarca) [inline frontend/src/components/widget/Attributes/Attributes.js:125]: pls export the constants.
pls document what's the meaning of 13, 400 and 150.

File: frontend/src/components/table/TableCell.js [hunk 4]
```diff
@@ -437,6 +441,8 @@ TableCell.propTypes = {
   modalVisible: PropTypes.bool,
   docId: PropTypes.any,
   activeLocale: PropTypes.object,
+  updateHeight: PropTypes.func,
```
Comment (teosarca) [inline frontend/src/components/table/TableCell.js:444]: pls document what's the purpose of updateHeight

File: frontend/src/components/table/TableCell.js [hunk 5]
```diff
@@ -437,6 +441,8 @@ TableCell.propTypes = {
...
+  arrIndex: PropTypes.number,
```
Comment (teosarca) [inline frontend/src/components/table/TableCell.js:445]: pls document what's the purpose of arrIndex.

File: frontend/src/components/widget/Attributes/Attributes.js [hunk 6]
```diff
@@ -359,5 +375,7 @@ Attributes.propTypes = {
   viewId: PropTypes.any,
   fieldName: PropTypes.any,
   entity: PropTypes.any,
+  updateHeight: PropTypes.func, // adjusts the table container with a given height from a child component when child exceeds visible area
+  arrIndex: PropTypes.number, // used for knowing the row index within the Table (used on AttributesDropdown component)
```
Comment (teosarca) [inline frontend/src/components/widget/Attributes/Attributes.js:379]: ok but then why not calling it rowIndex?

File: frontend/src/components/table/TableCell.js [hunk 7]
```diff
@@ -437,6 +441,8 @@ TableCell.propTypes = {
   modalVisible: PropTypes.bool,
   docId: PropTypes.any,
   activeLocale: PropTypes.object,
+  updateHeight: PropTypes.func, // adjusts the table container with a given height from a child component when child exceeds visible area
+  arrIndex: PropTypes.number, // used for knowing the row index within the Table (used on AttributesDropdown component)
```
Comment (teosarca) [inline frontend/src/components/table/TableCell.js:445]: why not naming it `rowIndex`?

File: frontend/src/components/table/TableItem.js [hunk 8]
```diff
@@ -733,6 +737,8 @@ TableItem.propTypes = {
   page: PropTypes.number,
   activeSort: PropTypes.bool,
   activeLocale: PropTypes.object,
+  updateHeight: PropTypes.func,
+  arrIndex: PropTypes.number,
```
Comment (teosarca) [inline frontend/src/components/table/TableItem.js:741]: why not naming it `rowIndex`?

Comment (teosarca) [reply frontend/src/components/table/TableItem.js:741]: also pls document it

**Comment by siemiatj** [comment]
I would only treat this as a temporary solution as imho this will affect the performance forcing re-renders. If we can't flip the list to stay on top, then I guess we'll have to live with that for now.

Screenshot 2020-06-16 at 01 57 38

**Comment by petrican** [comment]
> I would only treat this as a temporary solution as imho this will affect the performance forcing re-renders. If we can't flip the list to stay on top, then I guess we'll have to live with that for now.
> 
> Screenshot 2020-06-16 at 01 57 38

@siemiatj this solution indeed is going to re-render table. I'm not happy with it.  Need to try that react-tether.

**Comment by petrican** [comment]
https://www.loom.com/share/47379f73bc3440c099b0fbcd006bf379

Ok so I tested with the react-tether as well and that would introduce some other issues (additional scrollbar on the right, calendar within the attributes dropdown will be displayed with an offset). After discussing with Kuba we decided to test my approach and see the exact impact on the re-renderings on our live (dev instance) with many records. If there's no impact on the performance we can proceed and merge this.

**Comment by siemiatj** [comment]
The only thing that comes to mind is that maybe we can attach the Tether component to the Table ? Not sure how it works right now but I'm almost sure by default it's just attached to the document body https://www.npmjs.com/package/react-tether#renderelementto--proptypesstring

**Comment by petrican** [comment]
Measured the performance with many rows and there is no degradation by using this solution.
Screenshot 2020-07-22 at 11 20 54
Merging

**Pull request merged**